### DISCUSSION TITLE
Authenticate ASP .Net Applications hosted in linux with gMSA

### DISCUSSION THREAD

**Discussion by IsuraD** [aws/credentials-fetcher#63] [Q&A] [answered]
Hi, 

I managed to obtain the Kerberos token using the credentials-fetcher.  Afterward, I tried to authenticate my .Net Application using the token.

I adjusted the KRB5CCNAME environment variable to use the newly generated cache in the grpc_cli call. Furthermore, I made updates to the realms and domain _realm parameters in the /etc/krb5.conf file.

Despite making these changes, I'm still encountering a 401 error.

```dbug: Microsoft.AspNetCore.Routing.EndpointRoutingMiddleware[1]
      Request matched endpoint 'MyMvcApp.Controllers.HomeController.Index (MyMvcApp)'
dbug: Microsoft.AspNetCore.Authentication.Negotiate.NegotiateHandler[9]
      AuthenticationScheme: Negotiate was not authenticated.
info: Microsoft.AspNetCore.Authorization.DefaultAuthorizationService[2]
      Authorization failed. These requirements were not met:
      DenyAnonymousAuthorizationRequirement: Requires an authenticated user.
```

When I follow-up the 401 as follows, "Server not found in Kerberos database"

```
curl --negotiate -u : http://localhost:5198/ -k -v http1.1
*   Trying 127.0.0.1:5198...
* Connected to localhost (127.0.0.1) port 5198 (#0)
* gss_init_sec_context() failed: Server not found in Kerberos database.
* Server auth using Negotiate with user ''
> GET / HTTP/1.1
> Host: localhost:5198
> User-Agent: curl/7.87.0
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 401 Unauthorized
< Content-Length: 0
< Date: Sun, 27 Aug 2023 17:26:08 GMT
< Server: Kestrel
* gss_init_sec_context() failed: Server not found in Kerberos database.
< WWW-Authenticate: Negotiate
```

Do you have any insights into what might have gone wrong?

**Comment by smhmhmd** [comment]
@IsuraD 
Kestrel needs keytab as mentioned [here](https://learn.microsoft.com/en-us/aspnet/core/security/authentication/windowsauth?view=aspnetcore-7.0&tabs=visual-studio#linux-and-macos-environment-configuration), kerberos ticket is not sufficient.
credentials-fetcher does not have keytab support at the moment.
You can create a keytab using adutil as mentioned [here](https://learn.microsoft.com/en-us/sql/linux/sql-server-linux-ad-auth-adutil-tutorial?view=sql-server-ver16#adutil-spn)

**Reply by IsuraD** [reply]
Is keytab file required in gSMA as the passwords are managed by AD? I am wondering which password is to use for keytab generation.

**Comment by smhmhmd** [comment]
If you domain-join and point the krb5.conf to /etc/krb5.keytab, your kestrel app will work using the machine principal.
However, if you want to use the gMSA principal, keytab support is required in credentials-fetcher.

**Reply by IsuraD** [reply]
Thans for the responses. Actually I tried that as well. I can see the machine principal entry in keytab file. 
```

klist
Ticket cache: KCM:0
Default principal: 192-168-1-105$@[CONTOSO.COM](http://contoso.com/)

Valid starting     Expires            Service principal
28/08/23 12:22:00  28/08/23 22:22:00  krbtgt/[CONTOSO.COM@CONTOSO.COM](mailto:CONTOSO.COM@CONTOSO.COM)
renew until 04/09/23 12:22:00
28/08/23 12:22:00  28/08/23 22:22:00  ldap/dc1.contoso.com@
renew until 04/09/23 12:22:00
Ticket server: ldap/[dc1.contoso.com@CONTOSO.COM](mailto:dc1.contoso.com@CONTOSO.COM)


 klist -k /etc/krb5.keytab
Keytab name: FILE:/etc/krb5.keytab
KVNO Principal

```

However, still I'm getting this error. 

```
curl --negotiate -u : http://localhost:5198/ -k -v --http1.1
*   Trying 127.0.0.1:5198...
* Connected to localhost (127.0.0.1) port 5198 (#0)
* gss_init_sec_context() failed: SPNEGO cannot find mechanisms to negotiate.
* Server auth using Negotiate with user ''
> GET / HTTP/1.1
> Host: localhost:5198
> User-Agent: curl/8.0.1
> Accept: */*
>
< HTTP/1.1 401 Unauthorized
< Content-Length: 0
< Date: Mon, 28 Aug 2023 02:24:07 GMT
< Server: Kestrel
* gss_init_sec_context() failed: SPNEGO cannot find mechanisms to negotiate.
< WWW-Authenticate: Negotiate
```

**Reply by smhmhmd** [reply]
@IsuraD

> However, still I'm getting this error.

The /etc/krb5.keytab file is usually under root privilege without read access to users.
For your sample test (not for production), you can change the privileges.
Or you can run the dotnet program under root.

**Reply by IsuraD** [reply]
Hi @smhmhmd , I am running the dotnet program as the root.
I suspect the SPNs are not properly validated for the machine principal.

**Reply by smhmhmd** [reply]
@IsuraD 
I have a writeup that we did internally some time back,  I cannot share here in github.
It was intended for a blog to use kestrel with windows auth, we did not publish it.
However, if you reach out to me at samiull at amazon.com, we can probably find a way to share with you.

**Reply by smhmhmd** [reply, marked as answer]
@IsuraD 
Here are steps to authenticate a dotnet kestrel-based app using windows authentication
```

•	Active Directory domain created in AWS Managed Active Directory ( you can also self-host)
•	Domain-join one Linux and one Windows instance
•	In Linux instance, do the following: 
o	Install latest dotnet
o	git clone https://github.com/aspnet/AspNetCore.git
o	cd AspNetCore/src/Security/Authentication/Negotiate/samples/NegotiateAuthSample
o	Edit the Properties/launchSettings.json file, for example, ‘ec2amaz-kaqvuf’ is the hostname of the domain-joined Linux instance. 
         **Please make sure to add your hostname instead**.
	diff --git a/src/Security/Authentication/Negotiate/samples/NegotiateAuthSample/Properties/launchSettings.json b/src/Security/Authentication/Negotiate/samples/NegotiateAuthSample/Properties/launchSettings.json
	index 9c6ea2ebb2..86b42ef455 100644
	--- a/src/Security/Authentication/Negotiate/samples/NegotiateAuthSample/Properties/launchSettings.json
	+++ b/src/Security/Authentication/Negotiate/samples/NegotiateAuthSample/Properties/launchSettings.json
	@@ -15,7 +15,7 @@
	         "ASPNETCORE_SERVER": "Kestrel",
	         "ASPNETCORE_ENVIRONMENT": "Development"
	       },
	-      "applicationUrl": "[https://localhost:5001;http://localhost:5000](https://localhost:5001%3Bhttp/localhost:5000)"
	+      "applicationUrl": "http://ec2amaz-kaqvuf:5050/"
	     },
	     "IIS Express": {
	       "commandName": "IISExpress",
	@@ -24,15 +24,6 @@
	         "ASPNETCORE_SERVER": "IISOutOfProc",
	         "ASPNETCORE_ENVIRONMENT": "Development"
	       }
	-    },
	-    "HttpSys": {
	-      "commandName": "Project",
	-      "launchBrowser": true,
	-      "environmentVariables": {
	-        "ASPNETCORE_SERVER": "HttpSys",
	-        "ASPNETCORE_ENVIRONMENT": "Development"
	-      },
	-      "applicationUrl": "http://localhost:5000/"
	     }
	   }
	}
o	‘sudo dotnet run’ in Linux
o	Login to domain-joined windows instance using domain-credentials like [username@contoso.com](mailto:username@contoso.com). 
	Because we’re testing from Windows Server we have to disable a function called IEESC.
	IEESC prevents Internet Explorer from automatically sending credentials. Use Server
	Manager to turn it off.
o	Paste ‘http://ec2amaz-kaqvuf:5050’ in the internet explorer browser, you should see ‘Authenticated? True …’ 
	Under ‘sudo dotnet run’ you should see logs of successful SPNEGO negotiate.
```

**Comment by saikiranakula-amzn** [comment]
Looking at the issue, we would like to know a few more details:

Does your container have access to AD domain?
Was the ticket used for communication with a specific service (such as SQL server)? If so, the service should be domain-joined with AD and the gMSA service account principal should have the privilege to access the service.

**Reply by IsuraD** [reply]
Hi @saikiranakula-amzn , 

Yes. I'm testing in a fedora VM. It has access to the AD.  It's just a sample MVC .Net application that doesn't have any external communucation.
